whois-api/src/main/java/net/ripe/db/whois/api/rdap/RdapRelationService.java

```diff
@@ -0,0 +1,200 @@
+                                      final IpEntry mostSpecificResource,
+                                      final Set<IpEntry> mostSpecificFillingOverlaps) {
+
+        mostSpecificFillingOverlaps.add(mostSpecificResource);
+
+        final IpInterval mostSpecificInterval = IpInterval.parse(mostSpecificResource.getKey().toString());
+        final List<IpEntry> parentList = ipTree.findFirstLessSpecific(mostSpecificInterval);
+
+        if (parentList.isEmpty()){
+            return;
+        }
+
+        final List<IpEntry> siblingsAndExact = findSiblingsAndExact(ipTree, parentList);
+
+        final IpInterval firstSibling = (IpInterval)siblingsAndExact.getFirst().getKey();
+        final IpInterval lastSibling = (IpInterval)siblingsAndExact.getLast().getKey();
+
+        final IpEntry parent = parentList.getFirst();
+        final IpInterval parentInterval = (IpInterval) parent.getKey();
+
+        if (!parentInterval.equals(searchIp) && // If the parent is already the search ip we stop
+                !childrenCoverParentRange(firstSibling, lastSibling, parentInterval)){
+            extractBottomMatches(ipTree, searchIp, parent, mostSpecificFillingOverlaps);
+        }
+    }
+
+    private static boolean childrenCoverParentRange(final IpInterval firstResource, final IpInterval lastResource, final IpInterval parent){
+        return firstResource.beginAsInetAddress().equals(parent.beginAsInetAddress()) &&
+                lastResource.endAsInetAddress().equals(parent.endAsInetAddress());
+    }
+
+    private static List<IpEntry> findSiblingsAndExact(final IpTree ipTree, final List<IpEntry> parent) {
+        return ipTree.findFirstMoreSpecific(IpInterval.parse(parent.getFirst().getKey().toString()));
+    }
+
+    private IpEntry searchUpResource(final IpTree ipTree, final IpInterval searchIp){
+        final List<IpEntry> parentList = ipTree.findFirstLessSpecific(searchIp);
+        if (parentList.isEmpty() || !existAndNoAdministrative(searchIp, parentList.getFirst())){
+            throw new RdapException("404 Not Found", "No up level object has been found for " + searchIp.toString(), HttpStatus.NOT_FOUND_404);
+        }
+        return parentList.getFirst();
+    }
+
+    private IpEntry searchTopLevelResource(final IpTree ipTree, final IpInterval searchIp){
+        try {
+            IpEntry ipEntry = searchUpResource(ipTree, searchIp);
+            return searchTopRemainingResource(ipTree, ipEntry);
+        } catch (RdapException ex){
+            throw new RdapException("404 Not Found", "No top-level object has been found for " + searchIp.toString(), HttpStatus.NOT_FOUND_404);
+        }
+    }
+
+    private IpEntry searchTopRemainingResource(final IpTree ipTree, IpEntry searchIp) {
+        try {
+            searchIp = searchUpResource(ipTree, (IpInterval) searchIp.getKey());
+            searchTopRemainingResource(ipTree, searchIp);
+        } catch (RdapException ex){
+            /*
```

Comment:
RdapException is very generic , I think you meant to check for  404 exception. Do you need to check status code as well?